### PR TITLE
Major LSP refactoring

### PR DESCRIPTION
**Pull request by MaxStalker**
There were multiple race-conditions between Language Client, Server, Monaco Editor and React state, which lead to a series of tricky to debug errors. 
- [x] Client/Server communication is moved into separate context for dissecting the problem and later ease of use
- [x] Automatically update code, when switching between editors to pickup latest changes in contracts and pass them to Language Server for verification
- [x] Refactored data handling, styles and components into separate files for better DevEx

Planned for next release:
- [ ] Updated Language Server package, which now does not cache

### CODE REVIEW

**Review by 10thfloor — APPROVED** [review]

File: src/components/CadenceEditor/index.tsx [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+  // TODO: restore view state in next implementation
+  /*
+  const saveEditorState = (id: string, viewState: any) => {
+    setEditorStates({
+      ...editorStates,
+      [id]: viewState,
+    });
+  };
+   */
```
Comment (10thfloor) [inline src/components/CadenceEditor/index.tsx:37]: Can we remove this?

File: src/components/CadenceEditor/index.tsx [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    // we will ignore text line, cause onChange is different for readme and other scripts
+    // @ts-ignore
+    project.active.onChange(editor.getValue());
+  }, 150);
+
+  useEffect(configureCadence, []);
+  useEffect(() => {
+    // TODO: save/restore view state with:
+    //  - use ref to track current active id
+    //  - const oldState = editor.saveViewState();
```
Comment (10thfloor) [inline src/components/CadenceEditor/index.tsx:98]: Is this still TODO?

File: src/components/CadenceEditor/index.tsx [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    if (editor) {
+      // Remove tracking of model updates to prevent re-rendering
+      if (editorOnChange.current) {
+        editorOnChange.current.dispose();
+      }
+
+      // To pick up new changes in accounts, we will track project's active item and then add and remove
+      // new line at EOF to trick Language Client to send code changes and reimport the latest changes,
+      // clearing errors and warning about missing fields.
```
Comment (10thfloor) [inline src/components/CadenceEditor/index.tsx:107]: Interesting hack. What if there is no newline to strip?

File: src/hooks/useLanguageServer.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      [languageServer]
+  )
+
+  useEffect(()=>{
+    if(languageServer){
+      languageServer.updateCodeGetter(getCode)
+    }
+  },[project.project.accounts])
+
+  // TODO: Disable this, once the cadence language server package is updated
```
Comment (10thfloor) [inline src/hooks/useLanguageServer.ts:106]: Still valid?

File: src/components/CadenceEditor/index.tsx (hunk 3 above)
Comment (MaxStalker) [reply src/components/CadenceEditor/index.tsx:107]: There always will be one - cause we are adding it before removing :)

**Comment by kerrywei** [comment]
can we provide a test plan in the PR description? such a big change deserves thorough testing

**Comment by MaxStalker** [comment]
> can we provide a test plan in the PR description? such a big change deserves thorough testing

That's a really good point. I will provide test plan, when I finish last two issues I've spotted:
- for some reason language server is taking old code to get a list of arguments
- sometimes we see that LS can't find model on the editor (probably some severed link during updated)

**Comment by MaxStalker** [comment]
Alright!
The issues should be resolved now :)

**Pull request merged**